### PR TITLE
InstallPlan refactor

### PR DESCRIPTION
**Pull request by ezyang**
Merged with HEAD. Continuation of #3609.

### CODE REVIEW

File: cabal-install/Distribution/Client/InstallPlan.hs [hunk 1]
```diff
@@ -641,3 +445,305 @@ configureInstallPlan solverPlan =
...
+
+    asReadyPackage (Configured pkg) = ReadyPackage pkg
+    asReadyPackage _ = error "InstallPlan.completed: internal error"
+
+failed :: GenericInstallPlan ipkg srcpkg
+       -> Processing -> UnitId
+       -> ([srcpkg], Processing)
+failed plan (Processing processingSet completedSet failedSet) pkgid =
+    assert (pkgv `IntSet.member` processingSet) $
+    assert (all (`IntSet.notMember` processingSet) (tail newlyFailed)) $
```
Comment (ezyang) [inline cabal-install/Distribution/Client/InstallPlan.hs]: I hate this, you're assuming that the first element returned by closure is the initial one, no reason to expect this to be the case from the docs. @dcoutts

**Comment by phadej** [comment]
Could you rebase so there aren't merge commit in between? Hard (impossible) to follow (and review) branch-y pull requests

**Comment by ezyang** [comment]
That might be a bit difficult, I'd have to reconstruct the merge from scratch and without testing the commits individually, the intermediary commits are unlikely to build. Maybe review Duncan's commits one-by-one, and then the merge/fixups all in one go?

File: cabal-install/Distribution/Client/InstallPlan.hs (hunk 1 above)
Comment (dcoutts) [inline cabal-install/Distribution/Client/InstallPlan.hs]: Previously the code used `Data.Graph.reachable` which does always have this behaviour. Suggestions welcome. You can see why we need the "others, not this one".

**Comment by dcoutts** [comment]
So I was also rebasing against master, and I've been going through and checking each intermediate one builds and tests pass, so perhaps I'll push that as a separate PR again and then diff it against this one and merge in anything @ezyang noticed that I missed.

**Comment by 23Skidoo** [comment]
LGTM, let's fix remaining failures and merge.

**Comment by dcoutts** [comment]
@23Skidoo so I've also finished rebasing now and I think @ezyang will be a little happier with it since it uses his Graph module more directly rather than bypassing and going direct to Data.Graph as @ezyang's own rebased version here does.

Will need to include windows fixes and anything else the build bots show up.

**Pull request closed without merging**